worldedit-core/src/main/java/com/fastasyncworldedit/core/util/task/KeyQueuedExecutorService.java

```diff
@@ -0,0 +1,239 @@
+    ) {
+        parent = new ThreadPoolExecutor(
+                corePoolSize,
+                maximumPoolSize,
+                keepAliveTime,
+                unit,
+                new LinkedBlockingQueue<>(),
+                threadFactory
+        );
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#shutdown()}
+     */
+    public void shutdown() {
+        parent.shutdown();
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#shutdownNow()}
+     */
+    @Nonnull
+    public List<Runnable> shutdownNow() {
+        return parent.shutdownNow();
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#isShutdown()}
+     */
+    public boolean isShutdown() {
+        return parent.isShutdown();
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#isTerminated()}
+     */
+    public boolean isTerminated() {
+        return parent.isTerminated();
+    }
+
+    /**
+     * Delegates to {@link ThreadPoolExecutor#awaitTermination(long, TimeUnit)}
+     */
+    public boolean awaitTermination(long timeout, @Nonnull TimeUnit unit) throws InterruptedException {
+        return parent.awaitTermination(timeout, unit);
+    }
+
+    protected <T> FutureTask<T> newTaskFor(Runnable runnable, T value) {
+        return new FutureTask<>(runnable, value);
+    }
+
+    protected <T> FutureTask<T> newTaskFor(Callable<T> callable) {
+        return new FutureTask<>(callable);
+    }
+
+    @Nonnull
+    public <T> Future<T> submit(@Nonnull K key, @Nonnull Callable<T> task) {
+        FutureTask<T> ftask = newTaskFor(task);
+        execute(key, ftask);
+        return ftask;
+    }
+
+    @Nonnull
+    public <T> Future<T> submit(@Nonnull K key, @Nonnull Runnable task, T result) {
+        FutureTask<T> ftask = newTaskFor(task, result);
+        execute(key, ftask);
+        return ftask;
+    }
+
+    @Nonnull
+    public Future<?> submit(@Nonnull K key, @Nonnull Runnable task) {
+        FutureTask<Void> ftask = newTaskFor(task, null);
+        execute(key, ftask);
+        return ftask;
+    }
+
+    public void execute(@Nonnull K key, @Nonnull FutureTask<?> command) {
+        KeyRunner runner;
+        synchronized (keyQueue) {
+            runner = keyQueue.merge(key, new KeyRunner(key), (existing, theNew) -> {
+                existing.add(command);
+                return existing;
+            });
+        }
+        runner.triggerRun();
+    }
+
+    private final class KeyRunner {
+
+        private final Queue<FutureTask<?>> tasks = new ConcurrentLinkedQueue<>();
+        private final K key;
+
+        private KeyRunner(K key) {
+            this.key = key;
+        }
+
+        void add(FutureTask<?> task) {
+            if (!tasks.add(task)) {
+                throw new RejectedExecutionException(rejection());
+            }
+        }
+
+        synchronized void triggerRun() {
+            Runnable task = tasks.remove();
```

Comment:
`remove()` throws an exception if the queue is empty, what's the intention here?